core/client/fs/src/main/java/alluxio/client/file/cache/LocalCacheFileInStream.java

```diff
@@ -0,0 +1,300 @@
+      PageId pageId = new PageId(mStatus.getFileId(), currentPage);
+      try (ReadableByteChannel cachedData = mCacheManager.get(pageId, currentPageOffset)) {
+        if (cachedData != null) { // cache hit
+          // wrap return byte array in a bytebuffer and set the pos/limit for the page read
+          ByteBuffer buf = ByteBuffer.wrap(b);
+          buf.position(off + bytesRead);
+          buf.limit(off + bytesRead + bytesLeftInPage);
+          // read data from cache
+          while (buf.position() != buf.limit()) {
+            if (cachedData.read(buf) == -1) {
+              break;
+            }
+          }
+          Preconditions.checkState(buf.position() == buf.limit());
+          bytesRead += bytesLeftInPage;
+          mPosition += bytesLeftInPage;
+        } else { // cache miss
+          byte[] page = readExternalPage(mPosition);
+          if (page.length > 0) {
+            mCacheManager.put(pageId, page);
+            System.arraycopy(page, currentPageOffset, b, off + bytesRead, bytesLeftInPage);
+            bytesRead += bytesLeftInPage;
+            mPosition += bytesLeftInPage;
+          }
+        }
+      }
+    }
+    Preconditions.checkState(bytesRead == len || (bytesRead < len && remaining() == 0),
+        "Invalid number of bytes read - "
+            + "bytes to read = %d, actual bytes read = %d, bytes remains in file %d",
+        len, bytesRead, remaining());
+    return bytesRead;
+  }
+
+  @Override
+  public long skip(long n) {
+    checkIfClosed();
+    if (n <= 0) {
+      return 0;
+    }
+    long toSkip = Math.min(remaining(), n);
+    mPosition += toSkip;
+    return toSkip;
+  }
+
+  @Override
+  public void close() throws IOException {
+    mCloser.close();
+  }
+
+  @Override
+  public long remaining() {
+    return mEOF ? 0 : mStatus.getLength() - mPosition;
+  }
+
+  @Override
+  public int positionedRead(long pos, byte[] b, int off, int len) throws IOException {
+    Preconditions.checkArgument(len >= 0, "length should be non-negative");
+    Preconditions.checkArgument(off >= 0, "offset should be non-negative");
+    Preconditions.checkArgument(pos >= 0, "position should be non-negative");
+    if (len == 0) {
+      return 0;
+    }
+    if (pos < 0 || pos >= mStatus.getLength()) { // at end of file
+      return -1;
+    }
+    int bytesRead = 0;
+    long currentPosition = pos;
+    long lengthToRead = Math.min(len, mStatus.getLength() - pos);
+    // for each page, check if it is available in the cache
+    while (bytesRead < lengthToRead) {
+      long currentPage = currentPosition / mPageSize;
+      int currentPageOffset = (int) (currentPosition % mPageSize);
+      int bytesLeftInPage = (int) Math.min(mPageSize - currentPageOffset, lengthToRead - bytesRead);
+      PageId pageId = new PageId(mStatus.getFileId(), currentPage);
+      try (ReadableByteChannel cachedData = mCacheManager.get(pageId, currentPageOffset)) {
```

Comment:
how shall we handle IOException here when reading cache? Assume we shall fallback to external stream? Currently the IOE will be escalated.